### DISCUSSION TITLE
Random Ideas

### DISCUSSION THREAD

**Discussion by bwosborne2** [zwave-js/zwave-js#3985] [Ideas]
I have some ( likely nor well thought out) ideas and might be willing to try and implement with my limited programming experience. This thread is not meant as a criticism in any way. Jus ideas on how to possibly improve some user experience. As background, I worked a bit on the device database aspect of openHAB Z-Wave, assisting the developer.

1. The current process to add a new device to the database it not very user friendly or forgiving. I see 2 possible solutions
    a. Add some web form based New Device functionality to the current database browser site
    b. More conveniently, possibly include this functionality ad part of `zwavejs2mqtt`.

2, Many entries in the current device database do not meet the latest formatting standards. Perhaps a script could suggest changes to bring entries into compliance.

3. This project holds great promise for openHAB users. The current bingind does not support 700 Series controllers, S2, or Smart Start. It also tends to not properly delete nodes from the controller. Since the developer moved to a different country a year ago, he no longer has any Z-wave devices but still spends time approving database entries and manually exporting the database to GitHub. Providing an `zwavejs` based solution could free up his time. His enterprise customers use only Zigbee. His binding is developed somewhat independent of the openHAB project so he just deals with any "challenges" caused by the core maintainers. For example, the lead developer of OH told me I needed to wait 3 months before I could use a newer 700 Series device because they broke backward compatibility during a minor version release cycle. That drove me over here & to Home Assistant.

I envision possibly making use of the websockets server * possibly some Python code from the HA integration. There is a good community Python 3 program called HABApp that interfaces with the openHAB REST API. I am uncomfortable providing long term support to any openHAB solution though.

First phase may be to develop a websockets client  for `nodejs2mqtt` that can pass the proper Python 3 code through HABApp to openHAB. I am not sure how easy it would be to display some control panel in the Java-based openHAB frontend. Their lead developer likes everything under the control of their core developers. Using the REST API decouples this somewhat.

I have no idea if any of this fits in to any of your roadmaps. I am also assuming you could have a presence on the OH User forum for support as needed.

Any constructive thought or ideas are welcomed,

Bruce

**Comment by AlCalzone** [comment]
> The current process to add a new device to the database it not very user friendly or forgiving. I see 2 possible solutions
a. Add some web form based New Device functionality to the current database browser site

I only see this one. That functionality should not live in downstream projects. The main problem is that someone would have to step up and build the solution - I don't have the capacity to do that too. Also, when changes to the config file format are made, like https://github.com/zwave-js/node-zwave-js/pull/3772, this would have to be updated ASAP.

I've tried to make the editing process as easy as possible without a GUI, by providing a pre-configured hosted VSCode environment with all dependencies installed. This means you get auto-completion and syntax help for config files and can run the check scripts locally.

> Many entries in the current device database do not meet the latest formatting standards. Perhaps a script could suggest changes to bring entries into compliance.

@blhoward2 is working on this. Its a very extensive task though.

> This project holds great promise for openHAB users [...] I have no idea if any of this fits in to any of your roadmaps

I'm open to hosting a potential solution in this org, but since I neither use openHAB nor Java, this would need a maintainer like the config file UI you proposed. I'm here for questions and dev support though.
I think the main benefit of the [websocket server](https://github.com/zwave-js/zwave-js-server) is that it can easily be consumed by a language specific wrapper like https://github.com/zwave-js/ZWaveJS.NET for .NET, which was built in a few days.

**Reply by bwosborne2** [reply]
> I only see this one. That functionality should not live in downstream projects. The main problem is that someone would have to step up and build the solution - I don't have the capacity to do that too. Also, when changes to the config file format are made, like #3772, this would have to be updated ASAP.

The database browser site would be better to update. We have no control if somebody is running an older `zwavejs2mqtt` version. If ewe ignore being able to edit an existing entry that simplifies things until the database entries are updated.



> @blhoward2 is working on this. Its a very extensive task though.

Agreed! I spent about 4 months cleaning up the smaller openHAB database a couple of years ago.



> I'm open to hosting a potential solution in this org, but since I neither use openHAB nor Java, this would need a maintainer like the config file UI you proposed. I'm here for questions and dev support though.

I envision the code be written in Python. I do not know Java either. I figured I might be able to "stand on the shoulders" of what has already been developed for HA.  I no longer use (or want to use) openHAB either.

**Reply by blhoward2** [reply]
While I like the idea of being openHAB's zwave solution, I don't think we want to be responsible for maintaining the integration code for that project. Nor should it be bolted onto zwavejs2mqtt. We already offer a socket server, so I think someone from that project would need to handle writing the integration code to speak to it. They can then embed the zwavejs2mqtt UI to control it, and run that docker to run the socket server. We can't be in the business of maintaining the integration code too, especially in Python or, God forbid, Java.

**Reply by bwosborne2** [reply]
What do you think about a web for for new database device entries? I would expect it could output a GitHub PR to be manually reviewed & approved.

I suggested user forum presence because likely most issues would relate to the Z-wave portion just like on the HA site. It might be possible to get some help on integration support from the HABApp developer. He knows Python & ues openHAB & Z-Wave.

**Reply by blhoward2** [reply]
We'd talked before about this. As it stands you can click edit in the database, hit the period key for VS Code, and it will submit a PR. 

The issue is that this is more complex than you'd think. The user needs to be able to specify options, set the read only flag, etc. They still need to be intimately familiar with the style
Guide. And, more critically, how do they then correct things after we QC their PR? They'd need to be able to reload their work, make the corrections (while seeing our comments from GitHub), etc. At this point only a handful of people have ever submitted a device file and a substantial portion of them are abandoned when we ask them to clean up the file per our style guide or add all the options (not just the ones they need), so it hasn't been a high priority. 

If it's only going to setup the start of the file, we already do that. We can import the file 90% complete in most cases and they just need to cleanup the labels and such. A web interface would also need to account for us being able to auto-generate the starting file. 

I just don't see this as a high need, at least at the moment. 

Further, we're templating device files as we go and the goal is for them to all be templated. This is hard to do automatically.

**Comment by bwosborne2** [comment]
Fair enough. Just different than the web forms used by OH.

I will see if I can try to cobble together something for OH as time permits. I know knoe not to think about a new device form.

Thanks all.
